### PR TITLE
(WHO-1163) Add manual promotion UI

### PR DESCRIPTION
**Pull request by thrnio**
This implements the frontend (and a few small backend fixes) for manually promoting images through a pipeline. Here's a bunch of screenshots: https://imgur.com/a/uwncB

### CODE REVIEW

**Comment by brimworks** [comment]
Note that Sam Heutmaker is no longer at Puppet, so perhaps @baveltman would be a better person to review this?

File: assets/js/actions/PipelineActions.js [hunk 1]
```diff
@@ -464,4 +464,12 @@ export function removeMainPipelineStage() {
         });
     });
   });
-}
\ No newline at end of file
+}
+
+export function guessPipelineComponentType(component) {
+  if (("destinationContainerRepoDomain" in component) || ("destinationContainerRepoId" in component)) {
+    return "CopyToRepository";
```
Comment (baveltman) [inline assets/js/actions/PipelineActions.js]: @thrnio - please make these strings constants that can be referenced across multiple files. This way you can change them in only one location if ever necessary.

File: assets/js/actions/PipelineActions.js [hunk 2]
```diff
@@ -464,4 +464,12 @@ export function removeMainPipelineStage() {
         });
     });
   });
-}
\ No newline at end of file
+}
+
+export function guessPipelineComponentType(component) {
+  if (("destinationContainerRepoDomain" in component) || ("destinationContainerRepoId" in component)) {
```
Comment (baveltman) [inline assets/js/actions/PipelineActions.js]: @thrnio - do you care about the prototype chain here? if so it is better to use component.hasOwnProperty(...)

File: assets/js/pages/Pipeline.js [hunk 3]
```diff
@@ -82,20 +83,37 @@ export default class Pipeline extends Component {
   }
   renderPipeline() {
     if (!this.props.reposMap) return;
-    let repoContainerPipeline = this.props.reposMap[this.props.pipelineStore.pipeline.containerRepoId]
+    let repoContainerPipeline = this.props.reposMap[this.props.pipelineStore.pipeline.containerRepoId];
```
Comment (baveltman) [inline assets/js/pages/Pipeline.js]: @thrnio - are we guranteed to have the long prop chain this.props.pipelineStore.pipeline.containerRepoId and this.props.pipelineStore.pipeline.components? looks like a potential NPE waiting to happen. As such, I suggest you use NPE check and throw an error if the required prop is not found for some reason.

For instance:

```
let repoContainerPipeline = NPECheck(this.props, "pipelineStore/pipeline/containerRepoId", null);
if (!repoContainerPipeline) {
// handle error
}
```

File: assets/js/pages/Pipeline.js [hunk 4]
```diff
@@ -82,20 +83,37 @@ export default class Pipeline extends Component {
...
+    let pipelineComponents = this.props.pipelineStore.pipeline.components.slice();
+    let pipelineComponentsToRender = [];
+
+    while (pipelineComponents.length > 0) {
+      let nextComponent = pipelineComponents.shift();
+      let automatic = true;
+      if (this.context.actions.guessPipelineComponentType(nextComponent) === "ManualPromotionGate") {
```
Comment (baveltman) [inline assets/js/pages/Pipeline.js]: @thrnio - instead of using the magic string "ManualPromotionGate", use the constant you will define here.

**Comment by thrnio** [comment]
@brimworks I had to make some small changes to the Java code, do you mind reviewing those?

**Comment by thrnio** [comment]
This should be good to go.

**Pull request merged**